applications/FreeSurfaceApplication/custom_utilities/edgebased_levelset_auxiliary_utils.h

```diff
@@ -0,0 +1,458 @@
+                for(unsigned int i = 1; i<TDim+1; i++)
+                {
+                    if(distances[0]*distances[i]<=0) //if the edge is divided
+                    {
+                        double delta_d = fabs(distances[i]) + fabs(distances[0]);
+
+                        if(delta_d>1e-15)
+                        {
+                            double Ni = fabs(distances[0]) / delta_d;
+                            double N0 = fabs(distances[i]) / delta_d;
+
+                            noalias(coord_on_0) = N0 * geom[0].Coordinates();
+                            noalias(coord_on_0) += Ni * geom[i].Coordinates();
+                        }
+                        else
+                            noalias(coord_on_0) = geom[0].Coordinates();
+
+                        break;
+
+                    }
+                }
+
+                //now calculate the distance of all the nodes from the elemental free surface
+                for(unsigned int i = 0; i<TDim+1; i++)
+                {
+                    noalias(temp) = geom[i].Coordinates();
+                    noalias(temp) -= coord_on_0 ;
+
+                    double real_distance = 0.0;
+                    for(unsigned int k=0; k<TDim; k++)
+                        real_distance += temp[k]*grad_d[k];
+                    real_distance = fabs(real_distance);
+
+                    double& dist_i = geom[i].FastGetSolutionStepValue(rDistanceVar);
+                    if(real_distance < dist_i)
+                        dist_i = real_distance;
+
+                    //select the nodes for the computation of the distance
+                    geom[i].GetValue(IS_VISITED) = 1;
+                }
+
+            }
+
+
+
+        }
+
+        //loop on all nodes to treat correctly the case of the surface coinciding with a node
+        //copy rDistance var to GetValue database
+        array_1d<double,3> aux;
+        for(ModelPart::NodesContainerType::iterator it =  r_model_part.NodesBegin(); it !=r_model_part.NodesEnd(); it++)
+        {
+            if(fabs(it->GetValue(rDistanceVar)) < tol)
+            {
+                it->FastGetSolutionStepValue(rDistanceVar) = 0.0;
+                it->GetValue(IS_VISITED) = 1;
+            }
+        }
+
+        //compute the distance using the element based approach
+        BodyDistanceCalculationUtils util;
+        util.CalculateDistances<TDim>(r_model_part.Elements(),rDistanceVar,max_distance);
+
+        //finally change the sign to the distance as needed
+        for(ModelPart::NodesContainerType::iterator it =  r_model_part.NodesBegin(); it !=r_model_part.NodesEnd(); it++)
+        {
+            if(it->GetValue(rDistanceVar) < 0)
+                it->FastGetSolutionStepValue(rDistanceVar) = -it->FastGetSolutionStepValue(rDistanceVar);
+        }
+
+        //check if something is wrong
+        for(ModelPart::NodesContainerType::iterator it =  r_model_part.NodesBegin(); it !=r_model_part.NodesEnd(); it++)
```

Comment:
This can be parallel.

Reply:
Indeed, I think that you can introduce this check in the previous loop to avoid looping the nodes twice.